### ISSUE TITLE
Problem with Optional As Any Parameters

### ISSUE THREAD

**Issue by jpbro** [WaynePhillipsEA/twinbasic#753]
**Describe the bug**
I'm using a TLB that effectively declares the CreateFile API as follows:

```
    Public Declare Function CreateFile Lib "kernel32" (ByVal lpFileName As String, _
                               ByVal dwDesiredAccess As Long, _
                               ByVal dwShareMode As Long, _
                               Optional ByRef lpSecurityAttributes As Any = 0, _
                               ByVal dwCreationDisposition As Long, _
                               ByVal dwFlagsAndAttributes As Long, _ 
                               ByVal hTemplateFile As Long) As Long
```

If I subsequently try to call CreateFile without passing the optional parameter, I would expect it to default to 0. Instead I get the following error:

`unable to convert type 'Variant' to type 'Any'`

**To Reproduce**
1. Reference the following TLB in a tB project: [winU.zip](https://github.com/WaynePhillipsEA/twinbasic/files/8118446/winU.zip)
2. Add the following code: 

```
    Public Sub Main()
        CreateFile "C:\abc.txt", 0, 0, , 0, 0, 0
    End Sub
```

3. Notice the `unable to convert type 'Variant' to type 'Any'` error in the Problems list.

**Expected behavior**
I would expect the code to compile and use the default value of "0" for the Optional As Any parameter.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.14.31

**Comment by WaynePhillipsEA** [comment]
Just for my own reference, the TLB definition is
```
        long _stdcall CreateFile(
                        LPWSTR lpFileName, 
                        long dwDesiredAccess, 
                        long dwShareMode, 
                        [in, optional, defaultvalue(0)] void* lpSecurityAttributes, 
                        long dwCreationDisposition, 
                        long dwFlagsAndAttributes, 
                        long hTemplateFile);
```

**Comment by wqweto** [comment]
Optional must not be followed by non-optional params. Perhaps the error msg is misleading.

**Comment by WaynePhillipsEA** [comment]
> Optional must not be followed by non-optional params. Perhaps the error msg is misleading.

Except in type libraries apparently... which is news to me too!

**Comment by jpbro** [comment]
> > Optional must not be followed by non-optional params. Perhaps the error msg is misleading.
> 
> Except in type libraries apparently... which is news to me too!

Does that mean we can get non-Optional params after Optional ones in tB? Or is that one of my bad ideas™?

**Comment by bclothier** [comment]
FWIW, I was surprised too to see that this was possible as I thought it was disallowed by MIDL compiler, too. [Documentation on `optional` attribute](https://docs.microsoft.com/en-us/windows/win32/midl/optional)

**Comment by wqweto** [comment]
This would be mind-boggling. How would you call such beast (still using the optionalness i.e default value from typelib) I’m wondering.

**Comment by WaynePhillipsEA** [comment]
> FWIW, I was surprised too to see that this was possible as I thought it was disallowed by MIDL compiler, too.

That was exactly my thought too.

**Comment by jpbro** [comment]
@wqweto - I'm probably missing something, but all you would need are commas with no values/variables, no? e.g.: 

`CreateFile "C:\abc.txt", 0, 0, , 0, 0, 0  ' Skipped the 4th parameter so it uses the default value `

**Comment by WaynePhillipsEA** [comment]
> This would be mind-boggling. How would you call such beast (still using the optionalness i.e default value from typelib) I’m wondering.

See the OP...
`CreateFile "C:\abc.txt", 0, 0, , 0, 0, 0`

**Comment by bclothier** [comment]
Looks like. Not sure I have the old MKTYLEPLIB.

```
C:\temp\Test.idl(33) : warning MIDL2401 : [defaultvalue] is applied to a non-VARIANT and [optional]. Please remove [optional] : [ Parameter 'lpSecurityAttributes' of Procedure 'CreateFile' ( Module 'Test' ) ]
C:\temp\Test.idl(33) : error MIDL2090 : optional parameters must come after required parameters : [ Procedure 'CreateFile' ( Module 'Test' ) ]
C:\temp\Test.idl(33) : error MIDL2090 : optional parameters must come after required parameters : [ Procedure 'CreateFile' ( Module 'Test' ) ]
C:\temp\Test.idl(33) : error MIDL2090 : optional parameters must come after required parameters : [ Procedure 'CreateFile' ( Module 'Test' ) ]
```

**Comment by jpbro** [comment]
Anyway if it's illegal now, then forget I even asked ;) Although I could swear I've seen this "empty comma" skipping optional params approach elsewhere in VB-land, but I could be mistaken.

**Comment by WaynePhillipsEA** [comment]
@bclothier could you try the `/mktyplib203` MIDL switch?  (https://docs.microsoft.com/en-us/windows/win32/midl/-mktyplib203)

I'm not sure how accurate the switch is in simulating the exact MKTYPELIB behaviour.

**Comment by bclothier** [comment]
Even applying the switch, I still get the same warning & errors.

**Comment by WaynePhillipsEA** [comment]
> Anyway if it's illegal now, then forget I even asked ;) Although I could swear I've seen this "empty comma" skipping optional params approach elsewhere in VB-land, but I could be mistaken.

It seems like tB supports calling a function defined with these weird optionals... but it worries me that I have literally zero tests around that area, since I certainly didn't know it was possible/allowed.

This tB error looks to be to do with the default value on the `As Any`, not the weirdness of the optional parameter.

**Comment by bclothier** [comment]
I also wonder how one'd reflect it from the typelib API. IINM, we rely on the count of paramters - count of optionals.

**Comment by WaynePhillipsEA** [comment]
I'll try creating such a function through `ICreateTypeInfo` tomorrow.

**Comment by WaynePhillipsEA** [comment]
> I'll try creating such a function through `ICreateTypeInfo` tomorrow.

[ICreateTypeInfo::AddFuncDesc](https://docs.microsoft.com/en-us/windows/win32/api/oaidl/nf-oaidl-icreatetypeinfo-addfuncdesc) has no problem creating such a definition, so either this type library was generated programmatically, or it was generated from some other tool that doesn't have the same restrictions as MIDL.

**Comment by WaynePhillipsEA** [comment]
Fixed in v0.14.34, thanks!
